### ISSUE TITLE
3D plotting

### ISSUE THREAD

**Issue by Juanlu001** [poliastro/poliastro#22]
Summary:
- ~~Mayavi~~ https://github.com/enthought/mayavi is now having a second life, Python 3 support and Jupyter notebook exporting capabilities, but used to be buggy in the past
- ~~vispy~~ https://github.com/vispy/vispy seems to have stalled
- ~~PyOpenGL~~ http://pyopengl.sourceforge.net/ is a bit low level, seems more appropriate to render volumes
- ~~matplotlib~~ https://github.com/matplotlib/matplotlib has very limited 3D capabilities
- ~~Cesium~~ https://github.com/AnalyticalGraphicsInc/cesium seems more appropriate to render low-Earth orbits with realistic detail
- ~~GR~~ https://github.com/jheinen/gr does not work properly on Python 3 yet, does not solve matplotlib problems when used as a backend, is poorly documented, a tiny community and has lower quality results
- plotly https://github.com/plotly/plotly.py looks like the way to go!
- Bokeh just added extensions, which allow 3D plotting capabilities through JS libraries https://demo.bokehplots.com/apps/surface3d
- pythreejs https://github.com/jovyan/pythreejs

**Comment by Juanlu001** [comment]
Consider also vispy https://github.com/vispy/vispy

**Comment by Juanlu001** [comment]
vispy is a bit low level at this point, and forces you to use vanilla OpenGL for any non trivial work. Keep an eye on it but too early at the moment.

**Comment by Juanlu001** [comment]
New player: Glumpy https://glumpy.github.io/

**Comment by Juanlu001** [comment]
On second thought, for plotting lines and planets as point masses, matplotlib is enough. But then I have to distinguish the deep space case to actually paint the planets as points and not having to deal with matplotlib 3D limitations.

**Comment by Juanlu001** [comment]
And also: https://github.com/petrushy/CesiumWidget/ this is probably the most straightforward option, since it would suffice with a `poliastro -> CZML` conversion.

**Comment by Juanlu001** [comment]
I have just discovered https://github.com/sinhrks/cesiumpy, which is another Python wrapper for Cesium.js. I don't know if @petrushy and @sinhrks were aware of each other efforts, so I'll ping you both here in case you want to weigh in.

* CesiumWidget http://mybinder.org/repo/petrushy/CesiumWidget
* cesiumpy http://mybinder.org/repo/sinhrks/cesiumpy

**Comment by Juanlu001** [comment]
In any case, I don't even have to make a choice since as I stated before my work is to produce a `poliastro -> CZML` translation, which can later be read by either CesiumWidget or cesiumpy.

**Comment by sinhrks** [comment]
Thanks, I hadn't aware of CesiumWidget, and will take a look:)

**Comment by petrushy** [comment]
Hi! No, I wasn't aware of any other implementations, very interesting!

**Comment by Juanlu001** [comment]
Another contender: http://gr-framework.org/

**Comment by Juanlu001** [comment]
GR does not work properly on Python 3 yet, does not solve matplotlib problems when used as a backend, is poorly documented, a tiny community and has lower quality results.

**Comment by Juanlu001** [comment]
I created https://github.com/poliastro/poliastro/issues/117 as a separate issue to think about CZML exporting, and for this milestone I am focusing on getting 3D plotting with matplotlib right. Those not interested feel free to unsubscribe from this particular thread.

**Comment by Juanlu001** [comment]
I'm postponing this to 0.6 to not block the release.

**Comment by Juanlu001** [comment]
Thanks to @guillem-db I discovered that plotly has quite advanced 3D plotting capabilities, and respects the depth and opacity of objects. I will definitely have to check it out!

**Comment by Juanlu001** [comment]
Bokeh now has 3D! :scream: 

https://demo.bokehplots.com/apps/surface3d

(see https://github.com/bokeh/bokeh/issues/3693)

although "it's probably out of scope for the core library", it would be worth it to explore extensions @guillem-db.

**Comment by Juanlu001** [comment]
Another contender: https://github.com/jovyan/pythreejs

**Comment by Juanlu001** [comment]
Interestingly, advancing 3D plotting might be easier than refactoring 2D plotting since I don't have to convert to perifocal coordinates.

**Comment by Juanlu001** [comment]
ipywidgets 6.0 is released :tada::tada::tada:

http://blog.jupyter.org/2017/03/01/ipywidgets-6-release/

whatever we do for 3D visualization, it **must** be a Jupyter widget.

**Comment by Guillem-db** [comment]
Have you already thought something about how you want your widgets? I have a new Shaolin version compatible with ipywidgets 6. It is able to automagically build a dashboard directly from a kwargs dict.

**Comment by Juanlu001** [comment]
Hey @Guillem-db, thanks for chiming in! I'm not sure whether I will use something based on CesiumJS or roll my own thing, but in either case let us try and integrate it with Shaolin to provide a nice dashboard :D

**Comment by Juanlu001** [comment]
Progress and a survey of possible options here:

https://github.com/OpenDreamKit/OpenDreamKit/issues/86

**Comment by wonki22** [comment]
What about [http://vpython.org/](url)?
Here are some examples of what can be achieved with it [https://github.com/BruceSherwood/vpython-jupyter](url)
I think is an interesting option to consider.
Cons are it requires a different kernel in Jupyter and I don't know if that can affect too drastically to the project or if it's a minor issue. Also if you want to repeat an animation you need to restart the kernel which in my opinon is a little bit annoying but if so, at least you can cross it out of the candidates.

**Comment by Juanlu001** [comment]
The installation instructions recommend using the VPython kernel, but the contents of the `kernel.json` that [vpnotebook](https://pypi.python.org/pypi/vpnotebook/) installs are extremely simple:

```
{
 "language": "python",
 "argv": [
  "python",
  "-m",
  "ipykernel",
  "--InteractiveShellApp.exec_lines=['from __future__ import division, print_function','from math import *']",
  "-f",
  "{connection_file}"
 ],
 "display_name": "VPython"
}
```

I think everything should work if one does

```
from math import *
```

But this makes me think that it's not using the math functions of NumPy... which I find suspicious.

Apart from this, vpython-jupyter seems to support Python 3 and GlowScript apparently uses WebGL, so it might be a suitable alternative.

If you are considering working on it for SOCIS, start thinking of a work plan for the summer. I will send the instructions today or tomorrow to the mailing list :)

**Comment by Juanlu001** [comment]
This tutorial might be useful:

http://kazuar.github.io/jupyter-widget-tutorial/

**Comment by Juanlu001** [comment]
The sooner we accept we will have to learn JavaScript/TypeScript for this, the better... @anhiga

**Comment by anhiga** [comment]
Three.js specifically?

**Comment by Juanlu001** [comment]
> Three.js specifically?

Yes. There is pythreejs, but even so one has to write JS functions here and there, integrate it with the upcoming ipywidgets 7.0 machinery... There is no escape from it :)

**Comment by anhiga** [comment]
Well, knowing JS is a great skill nowadays, so... 😆

**Comment by Juanlu001** [comment]
I believe we should discuss #218 first, and therefore do not think we will make this in time for 0.7. Moving milestones.

**Comment by Juanlu001** [comment]
ipyvolume gained line plots :exclamation: 

https://ipyvolume.readthedocs.io/en/latest/api.html#ipyvolume.pylab.plot

Now we can avoid writing our own thing, at least for now @anhiga. We have a winner! :smile:

**Comment by Juanlu001** [comment]
...to clarify: the other winner has always been Cesium, of course. My concern has always been that it is too focused on rendering 3D globes and that it is too big. I went to the demo showcase again but found that the orbit predictor one is down:

http://cesiumjs.org/demos/OrbitalPredictor.html

http://www.orbitalpredictor.com/

Perhaps is worth a shot in any case.

**Comment by Juanlu001** [comment]
We don't have to get it right first time. Perhaps we can implement a basic version of both, and demand feedback from the community. I intend to work on this for 0.8.

**Comment by Juanlu001** [comment]
## Decision

Cesium is awesome, but perhaps _too complex_ for the purposes of poliastro. This projects has never had the intention of being a replacement of STK, but rather a tool for quickly experimentation and visualization of simple cases. We can implement CZML export anyway (see #117 for that).

On the other hand, although ipyvolume seems promising, the documentation is still lacking and the code difficult to navigate.

Although I've been reluctant to use plotly, it has a number of wins:

* Web native (as opposed to Mayavi)
* High level (pythreejs cannot say the same, let alone writing our own JavaScript/TypeScript)
* Mature (as opposed to ipyvolume)
* Science oriented (instead of data-oriented, like Bokeh and others)
* Self hosted (it wasn't when this issue was opened, see https://plot.ly/javascript/open-source-announcement/) and should be easy to configure for offline mode (see https://plot.ly/python/getting-started/#initialization-for-offline-plotting and https://www.reddit.com/r/IPython/comments/3tibc8/tip_on_how_to_run_plotly_examples_in_offline_mode/?ref=share&ref_source=link)
* Can export static images, see https://plot.ly/python/static-image-export/#export-static-image-offline (apparently Mayavi has issues with this, see https://github.com/OpenDreamKit/OpenDreamKit/issues/86#issuecomment-292076314)
* Can also work for 2D (and perhaps overcome some weird limitations of matplotlib layouts, see https://github.com/poliastro/poliastro/issues/205#issuecomment-327136362)
* Saw it used elsewhere for plotting orbits (nor SpiceyPy nor beyond, I can't remember now)
* Has a dedicated page on Aerospace uses, see https://plot.ly/products/industries/aerospace-analytics/

Example of 3D line plot in plotly:

https://plot.ly/python/3d-line-plots/

This issue is already long enough. Let's settle on this, and move away from plotly if we find serious limitations or awesome alternatives.

**Comment by Juanlu001** [comment]
Note: depending on time constraints, we might split this in two:

1) Provide a 3D plotting API, with a matplotlib backend (after #218, it should require few changes)
2) Change backend to Plotly

**Comment by Juanlu001** [comment]
Thanks to the new `.sample` method, 3D plotting got extremely easy:

![screenshot-2017-10-13 untitled](https://user-images.githubusercontent.com/316517/31534533-92766ae2-aff7-11e7-9656-911caf4920db.png)

The most complicated part there is... remembering how to create a 3D axes in matplotlib (I always have to look it up :joy:)

Plotting a globe is tempting, but we know the the results won't look good in matplotlib. This is totally uncoupled from the 2D simplification being tracked in #218 so I am unblocking this.
